programs/constrained_search/utils/sequential_threshold_processing.py

```diff
@@ -0,0 +1,440 @@
+"""Processes results files from multiple directories sequentially."""
+
+import argparse
+import glob
+import os
+from collections import defaultdict
+
+import numpy as np
+import pandas as pd
+from scipy.special import (
+    erfcinv,  # Required for the gaussian_noise_zscore_cutoff function
+)
+
+
+def gaussian_noise_zscore_cutoff(num_ccg: int, false_positives: float = 0.005) -> float:
+    """Determines the z-score cutoff based on Gaussian noise model and number of pixels.
+
+    NOTE: This procedure assumes that the z-scores (normalized maximum intensity
+    projections) are distributed according to a standard normal distribution. Here,
+    this model is used to find the cutoff value such that there is at most
+    'false_positives' number of false positives in all of the pixels.
+
+    Parameters
+    ----------
+    num_ccg : int
+        Total number of cross-correlograms calculated during template matching. Product
+        of the number of pixels, number of defocus values, and number of orientations.
+    false_positives : float, optional
+        Number of false positives to allow in the image (over all pixels). Default is
+        0.005 which corresponds to 0.5% false-positives.
+
+    Returns
+    -------
+    float
+        Z-score cutoff.
+    """
+    tmp = erfcinv(2.0 * false_positives / num_ccg)
+    tmp *= np.sqrt(2.0)
+
+    return float(tmp)
+
+
+def get_micrograph_id(filename: str) -> str:
+    """
+    Extract micrograph ID from filename.
+
+    Parameters
+    ----------
+    filename : str
+        Filename to extract micrograph ID from
+
+    Returns
+    -------
+    micrograph_id : str
+        Micrograph ID
+    """
+    base_name = os.path.basename(filename)
+    # Extract the part before _results.csv
+    parts = base_name.split("_results.csv")[0]
+    return parts
+
+
+def process_directories_sequentially(
+    directory_list: list[str],
```

Comment:
This function is fairly complex and difficult to follow what each part is doing. Should be split up into functional blocks (actually processing the data) and parts which print out some helpful information.